### ISSUE TITLE
Fog Volumes and Fog Shaders

### ISSUE THREAD

**Issue by clayjohn** [godotengine/godot-proposals#2261]
### Describe the project you are working on

Godot 4.0 :)

### Describe the problem or limitation you are having in your project
Right now volumetric fog uses a uniform density value for the entire view frustum. This constant density continues regardless of where the player is or what direction they are looking. This makes it impossible to have localized volumetric fog, volumetric fog that varies with height, fog that moves, swirls etc.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I propose we add a new node called ``FogVolume`` along with a new material ``FogVolumeMaterial`` and a new shader type ``shader_type fog``. 

The FogVolume is an Node that can be placed in the scene to add density to the volumetric fog. By default, it will be an AABB that takes a material. The material is responsible for setting the density over the volume of the AABB. 

The default FogVolumeMaterial, will have the following parameters:

- density (how much density to add to the area)
- edge fade (how much the density decreases at the edges, use this to get softer transitions)
- volume texture (3D texture that modulates the density)

Users can define custom shaders to do whatever they want. The fog shader will output two values: ``COLOR`` and ``DENSITY`` as input, it will have world space position, the extents of the AABB and the current ``UVW`` within the AABB (for texture lookups)

Importantly, we would need to add 2 new utilities:

1. An editor tool to convert a mesh into a 3D texture (this can work similar to how SDF particle colliders are generated, in fact SDF textures may be better than a straight voxelized texture)
2. A NoiseTexture3D resource

#### Optional

We may want to consider having a fog shader that runs on the entire view frustum each frame during the main density/lighting pass. This could be done as simply as injecting a fog shader into the main shader, the way we inject code into the default spatial/canvas_item shaders.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

*These are basic notes on implementation, I intend to add more detail and organization as I gain feedback on the core idea*

The way the Volumetric Fog works now, there is a full pass over the frustum to set the density value and light values. Then there is a couple passes to filter the density map. Finally, the density map is integrated to calculated the amount of fog at each position that can be used by the scene shader. 

This proposal requires us to inject density into the density map prior to filtering happening, likely even before the main density pass. 

Using the AABBs of the FogVolume node, we can easily compute an area of the frustum volume (density map) that may be touched and launch a compute shader for just that area. The compute shader could easily add density using atomic add functions. This would hopefully allow all fog volumes to write to the density map together. 

This approach is quite limited in that it doesnt allow for easy rotating of Volumes, rotations and transformations would have to be done in the shader (limited of course by the AABB)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No, this requires a new node type and changes to the volumetric fog shader

### Is there a reason why this should be core and not an add-on in the asset library?

It cannot be an addon.

**Comment by 2plus2makes5** [comment]
I don't know if it can help but here's a video on how fog works in Unreal Engine i saw some time ago:
https://www.youtube.com/watch?v=Xd7-rTzfmCo

**Comment by clayjohn** [comment]
@2plus2makes5 Thanks. I did look into Unreal's fog in researching this. In Unreal you write to fog using a special type of particle shader and an ExponentialHeightFog. I think overall the structure would be similar to my proposal, except in my proposal you have a single particle shader for a given area. 

But I guess since you mention it, maybe it does make sense to have a parent FogVolume and then any number of child FogParticles. The children could render to a 3D texture and then the Fog Volume could "stamp" that texture into the fog density buffer. 

This would allow a nice optimization where you could specify the resolution of the FogVolume (so low detail sections could be really cheap), and we could update the FogVolume only when changed or visible which would scale really well.

**Comment by LightningAA** [comment]
> A NoiseTexture3D resource

I'm probably misunderstanding what you mean, but doesn't [OpenSimplexNoise ](https://docs.godotengine.org/en/stable/classes/class_opensimplexnoise.html#class-opensimplexnoise-method-get-noise-3d) support 3d noise?

**Comment by clayjohn** [comment]
> I'm probably misunderstanding what you mean, but doesn't OpenSimplexNoise support 3d noise?

Yes. It does. But we don't have a NoiseTexture3D resource like we have a NoiseTexture2D. E.g. a resource that lets you generate an configure a noise texture in the editor

**Comment by clayjohn** [comment]
Doing some further research I have some ideas I want to jot down before incorporating them into the proposal:

1. We should add an ``anisotropy`` control to the volumetric fog controlled in the world environment (not on a fog volume basis)
2. We will need to run a full frustum density/color pass before the fog shaders run, this pass will reset the density to the value set in the world environment
        a. We should consider adding a "global fog shader" that is attached in the world environment and is run for every froxel, this will allow us to provide an easy to use material that does height based fog (based on world position) or noise based fog
3. We should consider exposing particle collision information to the fog shader
4. We should consider having the following structure:
        a. FogVolume: OOBB + material
        b. Fog Particle: child of FogVolume, sphere, has radius, position, density, color
        c. Internally, Godot will use the Fog Volume to carve out what area of the froxel buffer is impacted, and then launch a single compute dispatch to draw all child Particles. The particles can be stored in an array. The shader will loop through all the particles and run the same shader, adding the results together, and then finally stamping that into the froxel buffer.
        d. The benefit of this approach is that we only do one compute dispatch per fog volume, but we still have the flexibility to move particles around based on CPU computations (or maybe even based on a particle system in the future?).

**Comment by alexfreyre** [comment]
I think this idea is great and it will get used by almost everyone!  

Would be amazing if we can use any mesh as a volume or as @clayjohn suggested bake a mesh into a Texture3D

I agree with the possibility of adding a shader to the Volumetrig Fog so that everyone gets custom environments that can simulate dust or moisture in the air in some places and air movement if Texture3D is animated.

**Comment by Ansraer** [comment]
I haven't had time to read more about this, but from what I remember both Houdini and Blender support exporting fog/particles as openvdb files. Would be cool if we could import these files and use them as resources for fog volume density.

**Comment by fire** [comment]
If you have an openvdb sample I can try getting it into gltf for testing.

**Comment by Ophiolith** [comment]
Was just thinking about this while playing around with the current build of 4.0's volumetric fog and was considering submitting a proposal, but glad to see this is already being proposed. The fact that the current volumetric fog implementation is mostly limited to enabling 'godrays' and a single uniform volume of fog seems a little lacking. My ideas have more or less already been proposed by @clayjohn in a more eloquent manner than I could manage, but I can think of a few prominent existing examples of how what's proposed could be beneficial:

Warframe uses what I believe are froxel-based fog folumes with density modulated by a noise shader to give the impression of undulating fog in many areas of the game. Star Citizen uses what I think is a combination of procedural shaders for planetary cloud rendering, as well as predefined 3D textures for gas and nebula rendering. Many other games now use some kind of froxel-based volumetrics for rendering clouds, height-clamped fog etc.

This could of course also be beneficial in terms of bolstering Godot's existing PhysicalSky and ProceduralSky materials and could potentially eliminate the need to write separate shaders for cloud rendering as is current.

**Comment by Calinou** [comment]
> I think it would be cool to also have an option for fog volumes to cast shadows. You know, volumetric clouds that cast shadows and have effects on godrays.

Cloud shadows can be emulated well enough using Decals in 4.0 already :slightly_smiling_face:

**Comment by atirut-w** [comment]
> using the Decal node

How would you make them block godrays, though?

**Comment by Calinou** [comment]
> How would you make them block godrays, though?

You might be able to use lots of transparent surfaces arranged in planes, but it'll be slow.

**Comment by atirut-w** [comment]
> I would suggest just using fake volumetric lights using sprites at this point

So no realistic volumetric clouds that you can fly through and have them block godrays. :sob:

**Comment by clayjohn** [comment]
I started working on this. I figured it was pretty clear we needed some form of fog shader plus we would need to be able to do a full-frustum pass to do world-space effects. However, I am still uncertain how the API for fog _volumes_ should work. 

Accordingly, I will PR Fog shaders that includes a global, world-space pass, and then once we have feedback we can look into doing something like fog volumes or fog particles. 

I am certain that just having fog shaders will enable users to do some pretty nice things :)

![Screenshot (220)](https://user-images.githubusercontent.com/16521339/133192838-53684d9c-59ae-41bd-b1b5-99dc3f4cb364.png)

**Comment by clayjohn** [comment]
Implemented in https://github.com/godotengine/godot/pull/53353 :)

**Comment by macryc** [comment]
> I am certain that just having fog shaders will enable users to do some pretty nice things :)

You wouldn't mind sharing your fog () code for testing, would you?

**Comment by clayjohn** [comment]
> > I am certain that just having fog shaders will enable users to do some pretty nice things :)
> 
> You wouldn't mind sharing your fog () code for testing, would you?

Are you referring to the code in that screenshot? If so, unfortunately that code is long gone you can see most of it in the screenshot though! The noise function is generic 3D gradient noise (you can find many implementations on Shadertoy)

**Comment by lemilonkh** [comment]
> You wouldn't mind sharing your fog () code for testing, would you?

@macryc if you're still looking for this, I played around with it a bit and have a working example here: https://godotshaders.com/shader/moving-gradient-noise-fog-mist-for-godot-4

**Comment by fire** [comment]
Someone mentioned this seamless noise implmenetation. https://github.com/stegu/psrdnoise/tree/main
